engine/src/state.rs

```diff
@@ -10,9 +10,82 @@ pub use error::EngineStateError;
 const STATE_KEY: &[u8; 5] = b"STATE";
 
 /// Engine internal state, mostly configuration.
+#[derive(BorshSerialize, BorshDeserialize, Clone, PartialEq, Eq, Debug)]
+pub enum EngineState {
+    V2(EngineStateV2),
+    V1(EngineStateV1),
+}
```

Comment:
We don't necessarily have to do this, but I think this is a good opportunity to practice separation of concerns. By which I mean, right now `EngineState` has two distinct roles: it is used throughout our business logic to communicate config parameters but it is also used to represent the config information in the persistent storage. If we separate out `EngineState` (the object used in our business logic) from `BorshableEngineState` (the object used to represent the config in storage) then I think the code becomes simpler.

The issue with the current model of having one struct for two roles is that we need to write all these getter and setter methods to hide the complexity of `EngineState` being an enum from the business logic. In a world where they are separate we can write something like

```rust
/// Intentionally do not derive Borsh traits here because this type is storage-agnostic.
/// Note this type has the `bridge_prover_id` removed because it is not used in the business logic.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct EngineState {
    /// Chain id, according to the EIP-155 / ethereum-lists spec.
    pub chain_id: [u8; 32],
    /// Account which can upgrade this contract.
    /// Use empty to disable updatability.
    pub owner_id: AccountId,
    /// How many blocks after staging upgrade can deploy it.
    pub upgrade_delay_blocks: u64,
}

/// Intentionally make this type private to this module because the storage details do not matter
/// to the rest of the application.
#[derive(BorshSerialize, BorshDeserialize, Clone, PartialEq, Eq, Debug)]
enum BorshableEngineState {
    V2(EngineStateV2),
    V1(EngineStateV1),
}
```

Then the business logic remains unchanged (it still just see a struct) and we we can still do the state migration with backwards compatibility. Apart from making the code simpler, this approach should have a minor performance benefit over the current approach because we do not need all these `match` statements in all the getters and setters.

This is the approach I follow in the standalone engine for [TransactionMessage](https://github.com/aurora-is-near/aurora-engine/blob/develop/engine-standalone-storage/src/sync/types.rs#L30). You can see it is just a plain struct while [BorshableTransactionMessage](https://github.com/aurora-is-near/aurora-engine/blob/develop/engine-standalone-storage/src/sync/types.rs#L406) is an enum because of a prior state change.

Reply:
I like the idea of separating the `EngineState` 👍🏻